// apollo-server-core/dist/runQuery.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var graphql_1 = require("graphql");
var graphql_extensions_1 = require("graphql-extensions");
var apollo_tracing_1 = require("apollo-tracing");
var apollo_cache_control_1 = require("apollo-cache-control");
var errors_1 = require("./errors");
var logging_1 = require("./logging");
function runQuery(options) {
    return Promise.resolve().then(function () { return doRunQuery(options); });
}
exports.runQuery = runQuery;
function doRunQuery(options) {
    var documentAST;
    var logFunction = options.logFunction ||
        function () {
            return null;
        };
    var debugDefault = process.env.NODE_ENV !== 'production' && process.env.NODE_ENV !== 'test';
    var debug = options.debug !== undefined ? options.debug : debugDefault;
    logFunction({ action: logging_1.LogAction.request, step: logging_1.LogStep.start });
    var context = options.context || {};
    var extensions = [];
    if (options.tracing) {
        extensions.push(apollo_tracing_1.TracingExtension);
    }
    if (options.cacheControl === true) {
        extensions.push(apollo_cache_control_1.CacheControlExtension);
    }
    else if (options.cacheControl) {
        extensions.push(new apollo_cache_control_1.CacheControlExtension(options.cacheControl));
    }
    var extensionStack = extensions.length > 0 && new graphql_extensions_1.GraphQLExtensionStack(extensions);
    if (extensionStack) {
        context._extensionStack = extensionStack;
        graphql_extensions_1.enableGraphQLExtensions(options.schema);
        extensionStack.requestDidStart();
    }
    var qry = typeof options.query === 'string' ? options.query : graphql_1.print(options.query);
    logFunction({
        action: logging_1.LogAction.request,
        step: logging_1.LogStep.status,
        key: 'query',
        data: qry,
    });
    logFunction({
        action: logging_1.LogAction.request,
        step: logging_1.LogStep.status,
        key: 'variables',
        data: options.variables,
    });
    logFunction({
        action: logging_1.LogAction.request,
        step: logging_1.LogStep.status,
        key: 'operationName',
        data: options.operationName,
    });
    if (typeof options.query === 'string') {
        try {
            logFunction({ action: logging_1.LogAction.parse, step: logging_1.LogStep.start });
            documentAST = graphql_1.parse(options.query);
            logFunction({ action: logging_1.LogAction.parse, step: logging_1.LogStep.end });
        }
        catch (syntaxError) {
            logFunction({ action: logging_1.LogAction.parse, step: logging_1.LogStep.end });
            return Promise.resolve({
                errors: errors_1.formatApolloErrors([
                    errors_1.fromGraphQLError(syntaxError, {
                        errorClass: errors_1.SyntaxError,
                    }),
                ], {
                    formatter: options.formatError,
                    debug: debug,
                }),
            });
        }
    }
    else {
        documentAST = options.query;
    }
    var rules = graphql_1.specifiedRules;
    if (options.validationRules) {
        rules = rules.concat(options.validationRules);
    }
    logFunction({ action: logging_1.LogAction.validation, step: logging_1.LogStep.start });
    var validationErrors = graphql_1.validate(options.schema, documentAST, rules);
    logFunction({ action: logging_1.LogAction.validation, step: logging_1.LogStep.end });
    if (validationErrors.length) {
        return Promise.resolve({
            errors: errors_1.formatApolloErrors(validationErrors.map(function (err) {
                return errors_1.fromGraphQLError(err, { errorClass: errors_1.ValidationError });
            }), {
                formatter: options.formatError,
                logFunction: logFunction,
                debug: debug,
            }),
        });
    }
    if (extensionStack) {
        extensionStack.executionDidStart();
    }
    try {
        logFunction({ action: logging_1.LogAction.execute, step: logging_1.LogStep.start });
        return Promise.resolve(graphql_1.execute(options.schema, documentAST, options.rootValue, context, options.variables, options.operationName, options.fieldResolver)).then(function (result) {
            logFunction({ action: logging_1.LogAction.execute, step: logging_1.LogStep.end });
            var response = {
                data: result.data,
            };
            if (result.errors) {
                response.errors = errors_1.formatApolloErrors(result.errors.slice(), {
                    formatter: options.formatError,
                    logFunction: logFunction,
                    debug: debug,
                });
            }
            if (extensionStack) {
                extensionStack.executionDidEnd();
                extensionStack.requestDidEnd();
                response.extensions = extensionStack.format();
            }
            if (options.formatResponse) {
                response = options.formatResponse(response, options);
            }
            logFunction({
                action: logging_1.LogAction.request,
                step: logging_1.LogStep.end,
                key: 'response',
                data: response,
            });
            return response;
        });
    }
    catch (executionError) {
        logFunction({ action: logging_1.LogAction.execute, step: logging_1.LogStep.end });
        logFunction({ action: logging_1.LogAction.request, step: logging_1.LogStep.end });
        return Promise.resolve({
            errors: errors_1.formatApolloErrors([errors_1.fromGraphQLError(executionError)], {
                formatter: options.formatError,
                debug: debug,
            }),
        });
    }
}
